accepted/2024/support-nuget-authentication-plugins-dotnet-tools.md

```diff
@@ -0,0 +1,210 @@
+
+In the [current implementation](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs#L79-L101), the NuGet code searches for plugin files in a plugin-specific subdirectory. For example, in the case of the Azure Artifacts Credential Provider, the code looks for `*.exe` or `*.dll` files under the "CredentialProvider.Microsoft" directory. However, when .NET tools are installed in the `tools` folder, executable files (like `.exe` files on Windows or files with the executable bit set on Linux & Mac) are placed directly in the `tools` directory. The remaining files are stored in a `.store` folder. This arrangement eliminates the need to search in subdirectories.
+
+The new folder structure for NuGet plugins on the Windows platform is as follows: The `tools` folder contains .NET tool plugins, while the `.store` folder houses other necessary files.
+
+![plugin-tools-folder-windows](./../../meta/resources/PluginsAsDotNetTools/plugin-tools-folder-windows.png)
+
+![plugin-tools-store-windows](./../../meta/resources/PluginsAsDotNetTools/plugin-tools-store-windows.png)
+
+The new folder structure for NuGet plugins on the Linux platforms is as follows:
+
+```log
+{user}:~/.nuget/plugins$ dir
+netcore  tools
+
+{user}:~/.nuget/plugins$ cd tools/
+{user}:~/.nuget/plugins/tools$ dir
+botsay  dotnetsay
+
+{user}:~/.nuget/plugins/tools$ ls -la
+total 164
+drwxr-xr-x 3 {user}  4096 Feb 10 08:21 .
+drwxr-xr-x 4 {user}  4096 Feb 10 08:20 ..
+drwxr-xr-x 5 {user}  4096 Feb 10 08:21 .store
+-rwxr-xr-x 1 {user} 75632 Feb 10 08:21 botsay
+-rwxr-xr-x 1 {user} 75632 Feb 10 08:20 dotnetsay
+```
+
+## Drawbacks
+
+The ideal workflow for repositories accessing private NuGet feeds, such as Azure DevOps, involves installing the NuGet plugin as a .NET global tool. However, the current proposal suggests installing the plugin as a tool-path .NET tool. As mentioned in the technical explanation section, customers can opt to install NuGet plugins as a global tool instead of a tool-path tool. To do this, they need to set the `NUGET_DOTNET_TOOLS_PLUGIN_PATHS` environment variable. This variable should point to the location of the .NET Tool executable, which the NuGet Client tooling can invoke when needed.
+
+## Rationale and alternatives
+
+### NuGet commands to install credential providers
```

Comment:
just my 2 cents as the author of the `dotnet tool` proposal: we're definitely not tied to it explicitly, just the goal of having a simple, built in command, that brings down and makes credential providers usable for `dotnet restore` (or `nuget restore`) without any additional setup or configuration required. dotnet tool is one mechanism.

There's actually a benefit for using something like `dotnet nuget plugin` (or really something that goes through NuGet tooling) is that it could be made to work if using nuget.exe directly. Going though dotnet I would say is the mainline scenario, which is what I wanted to target, but there is a gap if users are using nuget.exe directly but without dotnet SDK installed; in this case the legacy credential provider install must be used (and something we'd likely have to support).

If installation and support is done by NuGet client tooling directly then a `dotnet nuget plugin` would work in the dotnet tooling, and conceivably support exposed via `nuget.exe plugin` as well.

Reply:
I don't think we have a concern for the nuget.exe provider. 

NuGet.exe *requires* VS to be installed, and VS itself ships with the NETFX version of the cred provider, so with NuGet.exe you shouldn't have a need to install the plugin.  It's there.

Reply:
> I think we could design a package and directory layout that might not be a huge effort to implement installing, updating, and uninstalling the plugins.

Thinking about it now, we wouldn't even need a new layout. 

I think the idea of "Launch the file that matches the folder name in the plugin folder" is still pretty sound. 
If we go new command, we can choose how many things we require to be different or how many things we break :)

Reply:
> NuGet.exe requires VS to be installed, and VS itself ships with the NETFX version of the cred provider, so with NuGet.exe you shouldn't have a need to install the plugin. It's there.

Thank you for adding this additional information, Nikolche. I was under the impression that NuGet.exe is a standalone tool that can run independently on Windows (& Mono with limited features) without requiring Visual Studio.

> The pros/cons list boils down to "it would be a better customer experience, but it would be more effort to implement". Why does this make the `dotnet tool` proposal the better choice? I guess it's implying that the amount of extra work is not worth the improved customer experience?

Thank you for the feedback, Andy. The goal is to design features that make it easy for customers to use them. My bias towards `.NET tools` is based on the following reasons:

- NuGet plugins are console applications. A `.NET tool` is a special NuGet package that contains a console application, which presents a natural fit.
- The .NET SDK has already simplified the process for customers to develop a tool. All the plugin authors have to do is set the following properties and execute `dotnet pack` to generate the package:

```xml
<PackAsTool>true</PackAsTool>
<ToolCommandName>botsay</ToolCommandName>
```

- Leveraging `.NET Tools` provides a standard experience for customers to share console applications, with NuGet plugins being one such use case. If there is an existing mechanism that works well, is familiar to customers, and fits our use case, I question the benefit of developing a new layout specifically for NuGet plugins.

- Another advantage of this established `.NET tools` approach is that plugin authors won't have to maintain separate code paths for .NET Framework and .NET Core runtimes. All these complexities are handled by the .NET SDK by providing a native shim upon the installation of the .NET tool.

Currently, at least to my understanding, the generated `.nupkg` will include a file named `DotnetToolSettings.xml` with additional metadata such as the command name, entry point, and runner.

For example, the [dotnetsay tool](https://nuget.info/packages/dotnetsay/2.1.7) includes the following content:

```xml
<?xml version="1.0" encoding="utf-8"?>
<DotNetCliTool Version="1">
  <Commands>
    <Command Name="dotnetsay" EntryPoint="dotnetsay.dll" Runner="dotnet" />
  </Commands>
</DotNetCliTool>
```

Considering all this, how about following a hybrid approach combining your design and the `.NET tools` approach?

- We will continue to use `.NET tools` as the mechanism for customers to develop the plugins.
- Instead of installing these `.NET tools` using the .NET SDK command, we will add `dotnet nuget plugin install/uninstall/update` commands so that customers don't have to remember the exact path where these tools need to be installed. As @baronfel mentioned in [NuGet/Home#13242 (comment)](https://github.com/NuGet/Home/pull/13242#discussion_r1489834403), these commands would act as wrappers on top of the existing `dotnet tool install/uninstall/update` commands by passing platform-specific parameters.  This approach will eliminate the need for customers to specify NuGet plugin paths based on their operating system.
- Following your [recommendation](https://github.com/NuGet/Home/issues/11325#issue-1031509172), we could also introduce a subcommand to search for packages with `packageType:CredentialProvider`.

If the team agrees with the proposed new design, I am happy to update the specification accordingly. In terms of a roadmap for implementation, I propose the following versions:

1. Initially, customers will install NuGet plugins using `.NET tools` commands by specifying the plugin's path based on their operating system. We will make the necessary code changes on the NuGet side to run the plugins installed as `.NET tools`.
2. Subsequently, we will introduce a `dotnet nuget plugin` subcommand to install, update, uninstall, and search for plugins. However, we will still rely on the `dotnet tool` command under the hood, as mentioned above. These command names are subject to change based on feedback to this specification, or we may create another specification just for those commands.

(I am still reviewing the other comments in this pull request; therefore, I wanted to document my proposal to ensure it is not forgotten. This proposal might change based on the other comments.)

Reply:
I couldn't agree more with these most recent comments about the pain points and extensibility. I too would love to see consideration for non-.NET providers, as many of the other systems I use in daily work lack a constraint like this, and I see no technical or protocol reasons why we should have such a restriction.

Reply:
> It's much better than the original proposal.

Thank you, Andy. Your comment has made me happy. I couldn’t have arrived at that conclusion without all of the great feedback from all the reviewers, and especially thank you for adding such a great detail in all your review comments.

I understand the limitations due to the lack of support for non-.NET providers. In my humble opinion (IMHO), we should create a new specification or work item to address this. Upon examining the code, I noticed that we launch these plugins in a separate process. It seems feasible to support plugins written in JavaScript, provided that the process has access to the Node.js runtime for executing the JavaScript code. As mentioned by @nkolev92 in this [discussion](https://github.com/NuGet/Home/pull/13242#discussion_r1491838316), developing a non-.NET provider requires more effort from the plugin authors because these non-.NET plugins/providers cannot directly access the NuGet SDK.

Reply:
> Additionally, from memory the original proposal didn't address discoverability at all either, not even a recommendation for cred providers to put the word NuGet in the package/tool name, so that dotnet tool search NuGet would allow customers to find credential providers.

I pushed a commit adding more information about the discoverability of credential providers at the end of the `Functional explanation` section in the spec. Please review and let me know your feedback.